lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
+  const SourceFile *SF;
+  const DependencyTracker &depTracker;
+  llvm::raw_fd_ostream &out;
+
+public:
+  using MemberTableEntryTy = std::pair<ReferencedNameTracker::MemberPair, bool>;
+
+  DependencyEmitter(const SourceFile *SF, const DependencyTracker &depTracker,
+                    llvm::raw_fd_ostream &out)
+      : SF(SF), depTracker(depTracker), out(out) {}
+
+  void emit() const;
+
+private:
+  void emitTopLevelNames(
+      ArrayRef<std::pair<DeclBaseName, bool>> sortedTopLevelNames) const;
+  void
+  emitMemberAccesses(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
+  void emitNominalTypes(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
+  void emitDynamicLookupMembers(const ArrayRef<std::pair<DeclBaseName, bool>>
+                                    sortedDynamicLookupNames) const;
+  void emitInterModuleReferences(const DependencyTracker &depTracker) const;
+};
+
+void ReferenceDependencyFileEmitter::emitDependencies() const {
+  DependencyEmitter(SF, depTracker, *out).emit();
 }
 
 bool swift::emitReferenceDependencies(DiagnosticEngine &diags, SourceFile *SF,
-                                      DependencyTracker &depTracker,
+                                      const DependencyTracker &depTracker,
                                       StringRef outputPath) {
-  assert(SF && "Cannot emit reference dependencies without a SourceFile");
-  
-  // Before writing to the dependencies file path, preserve any previous file
-  // that may have been there. No error handling -- this is just a nicety, it
-  // doesn't matter if it fails.
-  llvm::sys::fs::rename(outputPath, outputPath + "~");
-
-  std::error_code EC;
-  llvm::raw_fd_ostream out(outputPath, EC, llvm::sys::fs::F_None);
-
-  if (out.has_error() || EC) {
-    diags.diagnose(SourceLoc(), diag::error_opening_output, outputPath,
-                   EC.message());
-    out.clear_error();
-    return true;
+  if (auto rde = ReferenceDependencyFileEmitter::create(diags, SF, depTracker,
+                                                        outputPath)) {
+    rde->emit();
+    return false;
   }
+  return true;
+}
 
-  auto escape = [](DeclBaseName name) -> std::string {
-    return llvm::yaml::escape(name.userFacingName());
-  };
-
-  out << "### Swift dependencies file v0 ###\n";
+void ProviderEmitter::emit(const SourceFile *SF) {
+  emitTopLevelNames(SF->Decls);
+  emitNominalTypes();
+  emitMembers();
 
-  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
-  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
-  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+  if (SF->getASTContext().LangOpts.EnableObjCInterop)
+    emitDynamicLookupMembers(SF);
+}
 
-  out << "provides-top-level:\n";
-  for (const Decl *D : SF->Decls) {
-    switch (D->getKind()) {
-    case DeclKind::Module:
-      break;
+/// Write out the provided top level names and collect decls for later writing
+/// into three out-parameters.
```

Comment:
No longer accurate thanks to your nice refactor, although the function name doesn't quite describe the operation either…

Reply:
Removed the comment, which seems redundant. But I'm not sure what the issue is with the function name.

Reply:
Oh, just that `emitX` doesn't sound like it's also collecting state.